File: demo5/tools.py
from __future__ import annotations

import json
from pathlib import Path
from typing import Any

from config import GENERATED_FILES_DIR


def build_tools() -> list[dict[str, Any]]:
    """定义本节课可供模型自动选择的工具集合。"""
    return [
        {
            "type": "function",
            "function": {
                "name": "create_text_file",
                "description": (
                    "Create a text file under demo5/generated_files. "
                    "Use this when the user wants to save notes, outlines, plans, markdown, or code."
                ),
                "parameters": {
                    "type": "object",
                    "properties": {
                        "relative_path": {
                            "type": "string",
                            "description": "Relative file path under demo5/generated_files.",
                        },
                        "content": {
                            "type": "string",
                            "description": "Complete file content to write.",
                        },
                        "overwrite": {
                            "type": "boolean",
                            "description": "Whether to overwrite the file if it already exists.",
                        },
                    },
                    "required": ["relative_path", "content", "overwrite"],
                },
            },
        },
        {
            "type": "function",
            "function": {
                "name": "read_text_file",
                "description": (
                    "Read a text file under demo5/generated_files. "
                    "Use this when the user asks to inspect, verify, or review file content."
                ),
                "parameters": {
                    "type": "object",
                    "properties": {
                        "relative_path": {
                            "type": "string",
                            "description": "Relative file path under demo5/generated_files.",
                        }
                    },
                    "required": ["relative_path"],
                },
            },
        },
        {
            "type": "function",
            "function": {
                "name": "list_files",
                "description": (
                    "List files under demo5/generated_files. "
                    "Use this when the user asks what files exist or wants to inspect the directory."
                ),
                "parameters": {
                    "type": "object",
                    "properties": {
                        "relative_dir": {
                            "type": "string",
                            "description": "Relative directory under demo5/generated_files. Use '.' for the root.",
                        }
                    },
                    "required": ["relative_dir"],
                },
            },
        },
    ]


def resolve_safe_path(relative_path: str) -> Path:
    """把模型给出的相对路径解析成 demo5/generated_files 里的安全路径。"""
    cleaned_path = relative_path.strip().replace("\\", "/")
    if not cleaned_path:
        raise ValueError("relative_path 不能为空。")

    relative = Path(cleaned_path)
    if relative.is_absolute():
        raise ValueError("relative_path 不能是绝对路径。")

    target = (GENERATED_FILES_DIR / relative).resolve()
    base_dir = GENERATED_FILES_DIR.resolve()

    if target != base_dir and base_dir not in target.parents:
        raise ValueError("不允许访问 demo5/generated_files 目录之外的路径。")

    return target


def create_text_file(relative_path: str, content: str, overwrite: bool) -> dict[str, Any]:
    """创建文本文件。"""
    try:
        target_path = resolve_safe_path(relative_path)
    except ValueError as exc:
        return {"ok": False, "error": str(exc), "relative_path": relative_path}

    GENERATED_FILES_DIR.mkdir(parents=True, exist_ok=True)
    target_path.parent.mkdir(parents=True, exist_ok=True)

    existed_before = target_path.exists()
    if existed_before and not overwrite:
        return {
            "ok": False,
            "error": "文件已存在，且 overwrite 为 False。",
            "path": str(target_path),
        }

    target_path.write_text(content, encoding="utf-8")

    return {
        "ok": True,
        "path": str(target_path),
        "created": not existed_before,
        "overwritten": existed_before,
        "characters_written": len(content),
    }


def read_text_file(relative_path: str) -> dict[str, Any]:
    """读取文本文件。"""
    try:
        target_path = resolve_safe_path(relative_path)
    except ValueError as exc:
        return {"ok": False, "error": str(exc), "relative_path": relative_path}

    if not target_path.exists():
        return {
            "ok": False,
            "error": "文件不存在。",
            "path": str(target_path),
        }

    content = target_path.read_text(encoding="utf-8")
    return {
        "ok": True,
        "path": str(target_path),
        "content": content,
        "characters_read": len(content),
    }


def list_files(relative_dir: str) -> dict[str, Any]:
    """列出目录中的文件。"""
    try:
        target_dir = resolve_safe_path(relative_dir if relative_dir != "." else "")
    except ValueError as exc:
        return {"ok": False, "error": str(exc), "relative_dir": relative_dir}

    if not target_dir.exists():
        return {
            "ok": False,
            "error": "目录不存在。",
            "path": str(target_dir),
        }

    if not target_dir.is_dir():
        return {
            "ok": False,
            "error": "目标路径不是目录。",
            "path": str(target_dir),
        }

    items = []
    for item in sorted(target_dir.iterdir(), key=lambda p: p.name.lower()):
        items.append(
            {
                "name": item.name,
                "type": "dir" if item.is_dir() else "file",
            }
        )

    return {
        "ok": True,
        "path": str(target_dir),
        "items": items,
    }


def execute_tool_call(tool_call: dict[str, Any]) -> tuple[str, dict[str, Any]]:
    """
    执行模型返回的单个 tool_call。

    返回：
    - tool_name
    - tool_result
    """
    function_info = tool_call.get("function", {})
    tool_name = function_info.get("name", "unknown_tool")
    raw_arguments = function_info.get("arguments", "{}")

    try:
        arguments = json.loads(raw_arguments)
    except json.JSONDecodeError as exc:
        return tool_name, {
            "ok": False,
            "error": f"工具参数不是合法 JSON：{exc}",
            "raw_arguments": raw_arguments,
        }

    if tool_name == "create_text_file":
        return tool_name, create_text_file(
            relative_path=arguments["relative_path"],
            content=arguments["content"],
            overwrite=arguments["overwrite"],
        )

    if tool_name == "read_text_file":
        return tool_name, read_text_file(relative_path=arguments["relative_path"])

    if tool_name == "list_files":
        return tool_name, list_files(relative_dir=arguments["relative_dir"])

    return tool_name, {
        "ok": False,
        "error": f"未知工具：{tool_name}",
        "arguments": arguments,
    }

File: demo5/tools.js
import fsp from "node:fs/promises";
import path from "node:path";

import { GENERATED_FILES_DIR } from "./config.js";
import { resolveLikePython } from "../shared/safe_path.js";

// 定义本节课可供模型自动选择的工具集合。
export function buildTools() {
  return [
    {
      type: "function",
      function: {
        name: "create_text_file",
        description:
          "Create a text file under demo5/generated_files. " +
          "Use this when the user wants to save notes, outlines, plans, markdown, or code.",
        parameters: {
          type: "object",
          properties: {
            relative_path: {
              type: "string",
              description: "Relative file path under demo5/generated_files.",
            },
            content: {
              type: "string",
              description: "Complete file content to write.",
            },
            overwrite: {
              type: "boolean",
              description: "Whether to overwrite the file if it already exists.",
            },
          },
          required: ["relative_path", "content", "overwrite"],
        },
      },
    },
    {
      type: "function",
      function: {
        name: "read_text_file",
        description:
          "Read a text file under demo5/generated_files. " +
          "Use this when the user asks to inspect, verify, or review file content.",
        parameters: {
          type: "object",
          properties: {
            relative_path: {
              type: "string",
              description: "Relative file path under demo5/generated_files.",
            },
          },
          required: ["relative_path"],
        },
      },
    },
    {
      type: "function",
      function: {
        name: "list_files",
        description:
          "List files under demo5/generated_files. " +
          "Use this when the user asks what files exist or wants to inspect the directory.",
        parameters: {
          type: "object",
          properties: {
            relative_dir: {
              type: "string",
              description: "Relative directory under demo5/generated_files. Use '.' for the root.",
            },
          },
          required: ["relative_dir"],
        },
      },
    },
  ];
}

// 把模型给出的相对路径解析成 demo5/generated_files 里的安全路径。
async function resolveSafePath(relativePath) {
  const cleanedPath = String(relativePath ?? "").trim().replaceAll("\\", "/");
  if (!cleanedPath) {
    throw new Error("relative_path 不能为空。");
  }

  if (path.isAbsolute(cleanedPath)) {
    throw new Error("relative_path 不能是绝对路径。");
  }

  const baseDir = await resolveLikePython(GENERATED_FILES_DIR);
  const target = await resolveLikePython(path.join(GENERATED_FILES_DIR, cleanedPath));
  const relative = path.relative(baseDir, target);

  if (relative === "" || relative.startsWith("..") || path.isAbsolute(relative)) {
    throw new Error("不允许访问 demo5/generated_files 目录之外的路径。");
  }

  return target;
}

// 解析目录路径，'.' 表示 generated_files 根目录。
//
// 说明：Python 原版把 "." 转成空字符串传给 resolve_safe_path，
// 会因为“路径不能为空”直接报错（与工具描述“Use '.' for the root”矛盾），
// JS 版这里修正为返回根目录本身。
async function resolveSafeDir(relativeDir) {
  const cleaned = String(relativeDir ?? "").trim();
  if (cleaned === "" || cleaned === ".") {
    return resolveLikePython(GENERATED_FILES_DIR);
  }
  return resolveSafePath(relativeDir);
}

// 创建文本文件。
export async function createTextFile({ relative_path, content, overwrite }) {
  let targetPath;
  try {
    targetPath = await resolveSafePath(relative_path);
  } catch (error) {
    return { ok: false, error: error.message, relative_path };
  }

  await fsp.mkdir(path.dirname(targetPath), { recursive: true });

  let existedBefore = false;
  try {
    await fsp.access(targetPath);
    existedBefore = true;
  } catch {
    // 文件不存在
  }

  if (existedBefore && !overwrite) {
    return {
      ok: false,
      error: "文件已存在，且 overwrite 为 False。",
      path: targetPath,
    };
  }

  const contentString = String(content ?? "");
  await fsp.writeFile(targetPath, contentString, "utf-8");

  return {
    ok: true,
    path: targetPath,
    created: !existedBefore,
    overwritten: existedBefore,
    characters_written: contentString.length,
  };
}

// 读取文本文件。
export async function readTextFile({ relative_path }) {
  let targetPath;
  try {
    targetPath = await resolveSafePath(relative_path);
  } catch (error) {
    return { ok: false, error: error.message, relative_path };
  }

  let content;
  try {
    content = await fsp.readFile(targetPath, "utf-8");
  } catch (error) {
    if (error.code === "ENOENT") {
      return { ok: false, error: "文件不存在。", path: targetPath };
    }
    if (error.code === "EISDIR") {
      return { ok: false, error: "目标路径不是文件。", path: targetPath };
    }
    return { ok: false, error: `读取文件失败：${error.message}`, path: targetPath };
  }

  return {
    ok: true,
    path: targetPath,
    content,
    characters_read: content.length,
  };
}

// 列出目录中的文件。
export async function listFiles({ relative_dir }) {
  let targetDir;
  try {
    targetDir = await resolveSafeDir(relative_dir);
  } catch (error) {
    return { ok: false, error: error.message, relative_dir };
  }

  let entries;
  try {
    entries = await fsp.readdir(targetDir, { withFileTypes: true });
  } catch (error) {
    if (error.code === "ENOENT") {
      return { ok: false, error: "目录不存在。", path: targetDir };
    }
    if (error.code === "ENOTDIR") {
      return { ok: false, error: "目标路径不是目录。", path: targetDir };
    }
    return { ok: false, error: `读取目录失败：${error.message}`, path: targetDir };
  }

  const items = entries
    .map((entry) => ({
      name: entry.name,
      type: entry.isDirectory() ? "dir" : "file",
    }))
    .sort((a, b) => {
      const left = a.name.toLowerCase();
      const right = b.name.toLowerCase();
      if (left < right) return -1;
      if (left > right) return 1;
      return 0;
    });

  return {
    ok: true,
    path: targetDir,
    items,
  };
}

/**
 * 执行模型返回的单个 tool_call。
 *
 * 返回：
 * - toolName
 * - toolResult
 */
export async function executeToolCall(toolCall) {
  const functionInfo = toolCall.function || {};
  const toolName = functionInfo.name || "unknown_tool";
  const rawArguments = functionInfo.arguments ?? "{}";

  let arguments_ = {};
  try {
    arguments_ = JSON.parse(rawArguments || "{}");
  } catch (error) {
    return [
      toolName,
      {
        ok: false,
        error: `工具参数不是合法 JSON：${error.message}`,
        raw_arguments: rawArguments,
      },
    ];
  }

  if (toolName === "create_text_file") {
    const missing = ["relative_path", "content", "overwrite"].filter(
      (key) => arguments_[key] === undefined
    );
    if (missing.length > 0) {
      return [
        toolName,
        {
          ok: false,
          error: `缺少必要参数：${missing.join(", ")}`,
          arguments: arguments_,
        },
      ];
    }
    return [toolName, await createTextFile(arguments_)];
  }

  if (toolName === "read_text_file") {
    if (arguments_.relative_path === undefined) {
      return [
        toolName,
        {
          ok: false,
          error: "缺少必要参数：relative_path",
          arguments: arguments_,
        },
      ];
    }
    return [toolName, await readTextFile(arguments_)];
  }

  if (toolName === "list_files") {
    if (arguments_.relative_dir === undefined) {
      return [
        toolName,
        {
          ok: false,
          error: "缺少必要参数：relative_dir",
          arguments: arguments_,
        },
      ];
    }
    return [toolName, await listFiles(arguments_)];
  }

  return [
    toolName,
    {
      ok: false,
      error: `未知工具：${toolName}`,
      arguments: arguments_,
    },
  ];
}
